x-pack/plugins/proxy/server/cluster_doc.ts

```diff
@@ -0,0 +1,259 @@
+    this.timer = setTimeout(async () => {
+      this.log.debug('Updating node map');
+      await this.mainLoop();
+    }, this.updateInterval);
+  }
+
+  private updateRoutingTable(routingTable: RoutingTable): void {
+    const currentRoutes = [...Object.keys(this.routingTable)];
+    for (const [key, node] of Object.entries(routingTable)) {
+      this.routingTable[key] = node;
+      const idx = currentRoutes.findIndex(k => k === key);
+      if (idx) currentRoutes.splice(idx, 1);
+    }
+
+    for (const key of currentRoutes.values()) {
+      delete this.routingTable[key];
+    }
+  }
+
+  private async getNodeList(): Promise<NodeList> {
+    if (!this.elasticsearch) {
+      const err = Boom.boomify(new Error('You must call setup first'), { statusCode: 412 });
+      throw err;
+    }
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      _source: true,
+    };
+    const reply = await client.callAsInternalUser('get', params);
+    this.seq_no = reply._seq_no;
+    this.primary_term = reply._primary_term;
+    const data: ClusterDoc = reply._source;
+    this.updateRoutingTable(data.routing_table || {});
+    const nodes: NodeList = data.nodes || {};
+    return nodes;
+  }
+
+  private async updateNodeList(nodes: NodeList): Promise<void> {
+    if (!this.elasticsearch) {
+      const err = Boom.boomify(new Error('You must call setup first'), { statusCode: 412 });
+      throw err;
+    }
+    const doc = {
+      nodes,
+      routing_table: this.routingTable,
+    };
+    const client = await this.elasticsearch.pipe(first()).toPromise();
+    const params = {
+      id: this.proxyDoc,
+      index: this.proxyIndex,
+      if_seq_no: this.seq_no,
+      if_primary_term: this.primary_term,
+      body: doc,
+    };
+    await client.callAsInternalUser('index', params);
+  }
+
+  private updateLocalNode(nodes: NodeList, finishTime: number): NodeList {
+    nodes[this.nodeName] = {
+      lastUpdate: finishTime,
+    };
+    return nodes;
+  }
+
+  private removeNode(node: string) {
+    for (const [resource, data] of Object.entries(this.routingTable)) {
+      if (data.node === node) {
+        delete this.routingTable[resource];
+      }
+    }
+  }
+
+  private async mainLoop(): Promise<void> {
+    const nodes = await this.getNodeList();
+    const finishTime = new Date().getTime();
+
+    for (const [key, node] of Object.entries(nodes)) {
+      const timeout = finishTime - node.lastUpdate;
+      if (!node || timeout > this.timeoutThreshold) {
+        this.log.warn(`Node ${key} has not updated in ${timeout}ms and has been dropped`);
+        this.removeNode(key);
+        delete nodes[key];
+      }
+    }
+
+    try {
+      await this.updateNodeList(this.updateLocalNode(nodes, finishTime));
+    } catch (err) {
+      // on conflict, skip until next loop
+      if (err.output.statusCode !== 409) {
```

Comment:
that means if once non 409 error happen we are never going to recover?

it's possible es connection break at some time then recover.

I suggest when 409 happens, retry immediately, otherwise, retry using normal interval

Reply:
This is contrary to the consensus we agreed upon in the zoom call and via slack yesterday.

```
Todd Kennedy [1:18 PM]
OK
To summarize:
1. I'm going to ensure when the document update occurs it includes the version id of the last retrieved version as to prevent overwrites
2. if a 409 conflict is returned from es, we will ignore it and hope it resolves on the next cycle of the loop
3. this means that the proxy nodes can have an inconsistent state (lack consensus among peers) and we are _OK_ with that
```

Reply:
for 2, we agree we hope it resolve in next retry, but I didn't see how the retry would happen in this code?

Reply:
ah good point. subsequent runs of the cluster doc loop would cause errors to be swallowed since they're not being returned anywhere.  the cluster doc is now an event emitter and the proxy exports and interface to listen to these errors.

the consumer of the proxy can look at the error and determine if it should do something or let it side.

Reply:
update from this morning's meeting: this will continue to retry if it encounters and error, but during error state all proxy requests will fail